File: packages/sdk-java/src/main/java/org/melaya/TemplatesAPI.java
package org.melaya;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.Map;

/**
 * Templates API — create, manage, and share pipeline templates.
 *
 * <p>Maps to {@code /api/v1/private/user-templates/*} (user CRUD) and
 * {@code /api/v1/private/templates/*} (global / validated reads).
 *
 * <p>Visibility levels:
 * <ul>
 *   <li>{@code private} — only the creator</li>
 *   <li>{@code team} — members of the creator's projects</li>
 *   <li>{@code community} — all platform users</li>
 *   <li>{@code assigned} — explicitly assigned to users or projects</li>
 * </ul>
 *
 * @example
 * <pre>{@code
 * JsonNode templates = melaya.templates().list();
 * JsonNode created   = melaya.templates().save(Map.of(
 *     "name",    "My report",
 *     "payload", Map.of("steps", List.of("step1"))
 * ));
 * melaya.templates().share(created.get("id").asText(), "team");
 * }</pre>
 */
public class TemplatesAPI {

    private final HttpClient http;

    TemplatesAPI(HttpClient http) {
        this.http = http;
    }

    // ── User templates (CRUD) ─────────────────────────────────────────────────

    /**
     * List all templates visible to the caller — own + team + community +
     * assigned — filtered by Row-Level Security.
     */
    public JsonNode list() {
        return http.get("/api/v1/private/user-templates", null);
    }

    /**
     * Create a new private user template.
     *
     * @param body map containing {@code name} (required) and {@code payload} (required);
     *             optionally {@code description} and {@code category}
     */
    public JsonNode save(Map<String, Object> body) {
        return http.post("/api/v1/private/user-templates", body);
    }

    /**
     * Update name/description/category/payload of a private user template.
     *
     * @param templateId the template ID
     * @param body       fields to update
     */
    public JsonNode update(String templateId, Map<String, Object> body) {
        return http.patch("/api/v1/private/user-templates/" + encode(templateId), body);
    }

    /**
     * Duplicate a readable template into the caller's private library.
     *
     * @param templateId the source template ID
     */
    public JsonNode duplicate(String templateId) {
        return http.post(
                "/api/v1/private/user-templates/" + encode(templateId) + "/duplicate",
                null);
    }

    /**
     * Delete (or soft-demote if shared) a template.
     *
     * @param templateId the template ID
     */
    public JsonNode delete(String templateId) {
        return http.delete("/api/v1/private/user-templates/" + encode(templateId), null);
    }

    /**
     * Change the visibility of a template.
     *
     * @param templateId the template ID
     * @param visibility one of: {@code "private"}, {@code "team"}, {@code "community"}
     */
    public JsonNode share(String templateId, String visibility) {
        return http.put(
                "/api/v1/private/user-templates/" + encode(templateId) + "/visibility",
                MarketAPI.params("visibility", visibility));
    }

    // ── Assignments ───────────────────────────────────────────────────────────

    /**
     * List all assignments (users / projects) for a template.
     *
     * @param templateId the template ID
     */
    public JsonNode listAssignments(String templateId) {
        return http.get(
                "/api/v1/private/user-templates/" + encode(templateId) + "/assignments",
                null);
    }

    /**
     * Assign a template to a user or project.
     *
     * <p>Maps to {@code POST /api/v1/private/user-templates/:id/assignments}.
     * The server expects exactly one of {@code userId} or {@code projectId}
     * (both UUIDs) in the JSON body.
     *
     * @param templateId the template ID
     * @param target     map containing exactly one of {@code userId} or {@code projectId}
     *                   (UUID), e.g. {@code Map.of("userId", "...")} or
     *                   {@code Map.of("projectId", "...")}
     */
    public JsonNode assign(String templateId, Map<String, Object> target) {
        return http.post(
                "/api/v1/private/user-templates/" + encode(templateId) + "/assignments",
                target);
    }

    /**
     * Remove an assignment from a template.
     *
     * <p>Maps to {@code DELETE /api/v1/private/user-templates/:id/assignments}.
     * The server reads {@code userId}/{@code projectId} from <b>query params</b>
     * only — bodies on DELETE are ignored by the REST bridge.
     *
     * @param templateId the template ID
     * @param target     map containing exactly one of {@code userId} or {@code projectId}
     *                   (UUID) identifying the assignment to remove; sent as query params
     */
    public JsonNode unassign(String templateId, Map<String, Object> target) {
        return http.delete(
                "/api/v1/private/user-templates/" + encode(templateId) + "/assignments",
                target);
    }

    /** List projects the caller is a member of (for use in the share target picker). */
    public JsonNode shareTargets() {
        return http.get("/api/v1/private/user-templates/share-targets", null);
    }

    // ── Global / validated (read-only) ────────────────────────────────────────

    /** List all community-visibility (global) templates. */
    public JsonNode listGlobal() {
        return http.get("/api/v1/private/templates/global", null);
    }

    /** List IDs of all validated (platform-approved) templates. */
    public JsonNode listValidated() {
        return http.get("/api/v1/private/templates/validated", null);
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private static String encode(String s) {
        try {
            return java.net.URLEncoder.encode(s, java.nio.charset.StandardCharsets.UTF_8);
        } catch (Exception e) {
            return s;
        }
    }
}

File: packages/sdk-csharp/Melaya/TemplatesApi.cs
namespace Melaya;

/// <summary>
/// Templates API — create, manage, and share pipeline templates.
/// Maps to <c>/api/v1/private/user-templates/*</c> and <c>/api/v1/private/templates/*</c>.
/// </summary>
/// <example>
/// <code>
/// var templates = await m.Templates.ListAsync();
/// var tmpl = await m.Templates.SaveAsync(new TemplateSaveRequest { Name = "Daily report", Payload = ... });
/// await m.Templates.ShareAsync(tmpl.Id!, "team");
/// </code>
/// </example>
public sealed class TemplatesApi
{
    private readonly MelayaHttpClient _http;

    internal TemplatesApi(MelayaHttpClient http) => _http = http;

    /// <summary>List all templates visible to the caller — own + team + community + assigned (RLS-filtered).</summary>
    public async Task<List<UserTemplate>> ListAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<List<UserTemplate>>("/api/v1/private/user-templates", ct: ct).ConfigureAwait(false);
    }

    /// <summary>List all community-visibility (global) templates.</summary>
    public async Task<List<UserTemplate>> ListGlobalAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<List<UserTemplate>>("/api/v1/private/templates/global", ct: ct).ConfigureAwait(false);
    }

    /// <summary>List IDs of all validated (platform-approved) templates.</summary>
    public async Task<List<string>> ListValidatedAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<List<string>>("/api/v1/private/templates/validated", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Create a new private user template.</summary>
    public async Task<UserTemplate> SaveAsync(TemplateSaveRequest body, CancellationToken ct = default)
    {
        return await _http.PostAsync<UserTemplate>("/api/v1/private/user-templates", body, ct).ConfigureAwait(false);
    }

    /// <summary>Update name/description/category/payload of a private user template.</summary>
    public async Task<UserTemplate> UpdateAsync(string templateId, TemplateUpdateRequest body, CancellationToken ct = default)
    {
        return await _http.PatchAsync<UserTemplate>(
            $"/api/v1/private/user-templates/{Uri.EscapeDataString(templateId)}", body, ct).ConfigureAwait(false);
    }

    /// <summary>Duplicate a readable template into the caller's private library.</summary>
    public async Task<UserTemplate> DuplicateAsync(string templateId, string? newName = null, CancellationToken ct = default)
    {
        var body = newName is not null ? (object)new { newName } : new { };
        return await _http.PostAsync<UserTemplate>(
            $"/api/v1/private/user-templates/{Uri.EscapeDataString(templateId)}/duplicate", body, ct).ConfigureAwait(false);
    }

    /// <summary>Delete (or soft-demote if shared) a template.</summary>
    public async Task<BoolResult> DeleteAsync(string templateId, CancellationToken ct = default)
    {
        return await _http.DeleteAsync<BoolResult>(
            $"/api/v1/private/user-templates/{Uri.EscapeDataString(templateId)}", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Change the visibility of a template (<c>private</c>, <c>team</c>, <c>community</c>).</summary>
    public async Task<BoolResult> ShareAsync(string templateId, string visibility, CancellationToken ct = default)
    {
        var body = new { visibility };
        return await _http.PutAsync<BoolResult>(
            $"/api/v1/private/user-templates/{Uri.EscapeDataString(templateId)}/visibility", body, ct).ConfigureAwait(false);
    }

    // ── Assignments ───────────────────────────────────────────────────────────

    /// <summary>List all assignments (users / projects) for a template.</summary>
    public async Task<List<TemplateAssignment>> ListAssignmentsAsync(string templateId, CancellationToken ct = default)
    {
        return await _http.GetAsync<List<TemplateAssignment>>(
            $"/api/v1/private/user-templates/{Uri.EscapeDataString(templateId)}/assignments", ct: ct).ConfigureAwait(false);
    }

    /// <summary>
    /// Assign a template to a user or project.
    /// Set exactly one of <see cref="TemplateAssignRequest.UserId"/> or
    /// <see cref="TemplateAssignRequest.ProjectId"/>; the target is posted as
    /// <c>{ userId }</c> or <c>{ projectId }</c>.
    /// </summary>
    public async Task<BoolResult> AssignAsync(string templateId, TemplateAssignRequest body, CancellationToken ct = default)
    {
        ValidateTarget(body);
        return await _http.PostAsync<BoolResult>(
            $"/api/v1/private/user-templates/{Uri.EscapeDataString(templateId)}/assignments", body, ct).ConfigureAwait(false);
    }

    /// <summary>
    /// Remove an assignment from a template.
    /// Set exactly one of <see cref="TemplateAssignRequest.UserId"/> or
    /// <see cref="TemplateAssignRequest.ProjectId"/>; the target is sent as a
    /// <c>userId</c> / <c>projectId</c> query parameter because the server
    /// ignores DELETE request bodies.
    /// </summary>
    public async Task<BoolResult> UnassignAsync(string templateId, TemplateAssignRequest body, CancellationToken ct = default)
    {
        ValidateTarget(body);
        var query = new Dictionary<string, string?>
        {
            ["userId"]    = body.UserId,
            ["projectId"] = body.ProjectId,
        };
        return await _http.DeleteAsync<BoolResult>(
            $"/api/v1/private/user-templates/{Uri.EscapeDataString(templateId)}/assignments", query, ct).ConfigureAwait(false);
    }

    private static void ValidateTarget(TemplateAssignRequest body)
    {
        if ((body.UserId is null) == (body.ProjectId is null))
            throw new ArgumentException(
                "Melaya: set exactly one of `UserId` or `ProjectId` on TemplateAssignRequest.", nameof(body));
    }

    /// <summary>List projects the caller is a member of (for use in the share target picker).</summary>
    public async Task<List<ProjectShareTarget>> ShareTargetsAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<List<ProjectShareTarget>>("/api/v1/private/user-templates/share-targets", ct: ct).ConfigureAwait(false);
    }
}
